File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) 
    {
        int n=nums.length;
        int maxElement=-1;
        for(int num : nums)
        maxElement=Math.max(maxElement,num);

        int i=0;
        long freq=0,count=0;
        for(int j=0;j<n;j++)
        {
            if(nums[j]==maxElement)
            freq++;
            while(freq==k)
            {
                count+=n-j;       // subarrays starting from ith index.
                if(nums[i]==maxElement)
                freq--;
                i++;
            }
        }
        return count;
    }
}

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& a, int k) 
    {
        int maxi = *max_element(a.begin(),a.end());
        int n = a.size(), i = 0; 
        long long count = 0, freq = 0;
        for(int j = 0; j < n; j++)
        {
            if(a[j] == maxi)
            freq++;
            while(freq == k)
            {
                count += n - j;          // counting all subarrays starting from i.
                if(a[i] == maxi)
                freq--;
                i++;
            }
        }
        return count;
    }
};